File: internal/tsdb/ringbuffer.go
// Package tsdb provides an in-memory ring buffer for per-metric sliding windows
// with pre-computed aggregates. It accelerates dashboard queries for recent data
// so they never need to hit the relational DB.
package tsdb

import (
	"math"
	"sort"
	"sync"
	"time"
)

// WindowAgg is the pre-computed aggregate for one time window slot.
type WindowAgg struct {
	MetricName  string
	ServiceName string
	WindowStart time.Time
	Count       int64
	Sum         float64
	Min         float64
	Max         float64
	P50         float64
	P95         float64
	P99         float64
	samples     []float64 // raw samples, capped at maxSamples
}

const maxSamples = 256 // per-window sample cap for percentile computation

// ringSlot is a fixed-capacity circular slot in the ring buffer.
// Access is protected by the parent MetricRing's mu.
type ringSlot struct {
	agg WindowAgg
}

// MetricRing is a fixed-size circular buffer for a single metric+service key.
// Each slot represents one windowDuration-wide time bucket.
type MetricRing struct {
	mu           sync.RWMutex
	slots        []ringSlot
	size         int
	windowDur    time.Duration
	currentIdx   int
	currentStart time.Time
	metricName   string
	serviceName  string
}

func newMetricRing(metricName, serviceName string, slots int, windowDur time.Duration) *MetricRing {
	r := &MetricRing{
		slots:       make([]ringSlot, slots),
		size:        slots,
		windowDur:   windowDur,
		metricName:  metricName,
		serviceName: serviceName,
	}
	now := time.Now().UTC()
	r.currentStart = now.Truncate(windowDur)
	for i := range r.slots {
		r.slots[i].agg = newEmptyAgg(metricName, serviceName)
	}
	r.slots[0].agg.WindowStart = r.currentStart
	return r
}

func newEmptyAgg(metric, service string) WindowAgg {
	return WindowAgg{
		MetricName:  metric,
		ServiceName: service,
		Min:         math.MaxFloat64,
		Max:         -math.MaxFloat64,
	}
}

// Record adds a new data point, advancing the window if necessary.
func (r *MetricRing) Record(value float64, at time.Time) {
	r.mu.Lock()
	defer r.mu.Unlock()

	windowStart := at.UTC().Truncate(r.windowDur)
	if windowStart.After(r.currentStart) {
		// Advance one or more slots.
		steps := int(windowStart.Sub(r.currentStart) / r.windowDur)
		for i := 0; i < steps && i < r.size; i++ {
			r.currentIdx = (r.currentIdx + 1) % r.size
			r.currentStart = r.currentStart.Add(r.windowDur)
			s := &r.slots[r.currentIdx]
			s.agg = newEmptyAgg(r.metricName, r.serviceName)
			s.agg.WindowStart = r.currentStart
		}
	}

	s := &r.slots[r.currentIdx]
	s.agg.Count++
	s.agg.Sum += value
	if value < s.agg.Min {
		s.agg.Min = value
	}
	if value > s.agg.Max {
		s.agg.Max = value
	}
	if len(s.agg.samples) < maxSamples {
		s.agg.samples = append(s.agg.samples, value)
	}
}

// Windows returns up to `n` most-recent completed window aggregates.
// Takes a snapshot under the ring lock to avoid acquiring 120+ slot locks sequentially.
func (r *MetricRing) Windows(n int) []WindowAgg {
	r.mu.Lock()
	if n > r.size {
		n = r.size
	}
	// Snapshot all needed slots under a single lock
	type slotSnap struct {
		agg     WindowAgg
		samples []float64
	}
	snaps := make([]slotSnap, 0, n)
	for i := 0; i < n; i++ {
		idx := (r.currentIdx - i + r.size) % r.size
		s := &r.slots[idx]
		if s.agg.Count > 0 {
			samplesCopy := make([]float64, len(s.agg.samples))
			copy(samplesCopy, s.agg.samples)
			snaps = append(snaps, slotSnap{agg: s.agg, samples: samplesCopy})
		}
	}
	r.mu.Unlock()

	// Compute percentiles outside the lock
	out := make([]WindowAgg, 0, len(snaps))
	for _, snap := range snaps {
		agg := snap.agg
		agg.P50 = ringPercentile(snap.samples, 50)
		agg.P95 = ringPercentile(snap.samples, 95)
		agg.P99 = ringPercentile(snap.samples, 99)
		if agg.Min == math.MaxFloat64 {
			agg.Min = 0
		}
		if agg.Max == -math.MaxFloat64 {
			agg.Max = 0
		}
		agg.samples = nil // don't leak raw samples
		out = append(out, agg)
	}
	return out
}

// RingBuffer manages per-metric ring buffers, keyed by "service|metric".
type RingBuffer struct {
	mu        sync.RWMutex
	rings     map[string]*MetricRing
	slots     int
	windowDur time.Duration
}

// NewRingBuffer creates a RingBuffer.
// slots × windowDur = total retention (e.g. 120 × 30s = 1 hour).
func NewRingBuffer(slots int, windowDur time.Duration) *RingBuffer {
	return &RingBuffer{
		rings:     make(map[string]*MetricRing),
		slots:     slots,
		windowDur: windowDur,
	}
}

// Record records a value for the given metric+service at time t.
func (rb *RingBuffer) Record(metricName, serviceName string, value float64, at time.Time) {
	key := serviceName + "|" + metricName
	rb.mu.RLock()
	ring, ok := rb.rings[key]
	rb.mu.RUnlock()
	if !ok {
		rb.mu.Lock()
		ring, ok = rb.rings[key]
		if !ok {
			ring = newMetricRing(metricName, serviceName, rb.slots, rb.windowDur)
			rb.rings[key] = ring
		}
		rb.mu.Unlock()
	}
	ring.Record(value, at)
}

// QueryRecent returns aggregated windows for the given metric+service.
// Pass an empty serviceName to query across all services (returns first match).
func (rb *RingBuffer) QueryRecent(metricName, serviceName string, windowCount int) []WindowAgg {
	key := serviceName + "|" + metricName
	rb.mu.RLock()
	ring, ok := rb.rings[key]
	rb.mu.RUnlock()
	if !ok {
		return nil
	}
	return ring.Windows(windowCount)
}

// AllKeys returns all registered metric+service keys.
func (rb *RingBuffer) AllKeys() []string {
	rb.mu.RLock()
	defer rb.mu.RUnlock()
	keys := make([]string, 0, len(rb.rings))
	for k := range rb.rings {
		keys = append(keys, k)
	}
	return keys
}

// MetricCount returns the number of distinct metric series tracked.
func (rb *RingBuffer) MetricCount() int {
	rb.mu.RLock()
	defer rb.mu.RUnlock()
	return len(rb.rings)
}

// ringPercentile computes the p-th percentile of a float64 slice.
func ringPercentile(data []float64, p float64) float64 {
	if len(data) == 0 {
		return 0
	}
	sorted := make([]float64, len(data))
	copy(sorted, data)
	sort.Float64s(sorted)
	idx := int(math.Ceil(p/100*float64(len(sorted)))) - 1
	if idx < 0 {
		idx = 0
	}
	if idx >= len(sorted) {
		idx = len(sorted) - 1
	}
	return sorted[idx]
}

File: src/tsdb/ringbuffer.py
"""In-memory ring buffer for per-metric sliding windows with pre-computed aggregates."""

from __future__ import annotations

import math
import threading
import time as _time
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Dict, List, Optional


@dataclass
class WindowAgg:
    metric_name: str = ""
    service_name: str = ""
    window_start: float = 0.0  # unix timestamp
    count: int = 0
    sum: float = 0.0
    min: float = float("inf")
    max: float = float("-inf")
    p50: float = 0.0
    p95: float = 0.0
    p99: float = 0.0
    samples: list[float] = field(default_factory=list)

    def to_dict(self) -> dict:
        return {
            "metric_name": self.metric_name,
            "service_name": self.service_name,
            "window_start": datetime.fromtimestamp(self.window_start, tz=timezone.utc).isoformat()
            if self.window_start
            else None,
            "count": self.count,
            "sum": self.sum,
            "min": self.min if self.min != float("inf") else 0,
            "max": self.max if self.max != float("-inf") else 0,
            "p50": self.p50,
            "p95": self.p95,
            "p99": self.p99,
        }


MAX_SAMPLES = 256


def _percentile(data: list[float], p: float) -> float:
    if not data:
        return 0.0
    s = sorted(data)
    idx = int(math.ceil(p / 100.0 * len(s))) - 1
    idx = max(0, min(idx, len(s) - 1))
    return s[idx]


class MetricRing:
    """Fixed-size circular buffer for a single metric+service key."""

    def __init__(self, metric_name: str, service_name: str, slots: int, window_dur: float):
        self._lock = threading.Lock()
        self._slots: list[WindowAgg] = [
            WindowAgg(metric_name=metric_name, service_name=service_name) for _ in range(slots)
        ]
        self._size = slots
        self._window_dur = window_dur
        self._metric_name = metric_name
        self._service_name = service_name
        self._current_idx = 0
        now = _time.time()
        self._current_start = now - (now % window_dur)
        self._slots[0].window_start = self._current_start

    def record(self, value: float, at: float) -> None:
        with self._lock:
            window_start = at - (at % self._window_dur)
            if window_start > self._current_start:
                steps = int((window_start - self._current_start) / self._window_dur)
                for _ in range(min(steps, self._size)):
                    self._current_idx = (self._current_idx + 1) % self._size
                    self._current_start += self._window_dur
                    slot = self._slots[self._current_idx]
                    slot.count = 0
                    slot.sum = 0.0
                    slot.min = float("inf")
                    slot.max = float("-inf")
                    slot.samples = []
                    slot.window_start = self._current_start
                    slot.metric_name = self._metric_name
                    slot.service_name = self._service_name

            slot = self._slots[self._current_idx]
            slot.count += 1
            slot.sum += value
            if value < slot.min:
                slot.min = value
            if value > slot.max:
                slot.max = value
            if len(slot.samples) < MAX_SAMPLES:
                slot.samples.append(value)

    def windows(self, n: int) -> list[WindowAgg]:
        with self._lock:
            n = min(n, self._size)
            snapshots = []
            for i in range(n):
                idx = (self._current_idx - i + self._size) % self._size
                slot = self._slots[idx]
                if slot.count > 0:
                    agg = WindowAgg(
                        metric_name=slot.metric_name,
                        service_name=slot.service_name,
                        window_start=slot.window_start,
                        count=slot.count,
                        sum=slot.sum,
                        min=slot.min,
                        max=slot.max,
                        samples=list(slot.samples),
                    )
                    snapshots.append(agg)

        # Compute percentiles outside the lock
        result = []
        for agg in snapshots:
            agg.p50 = _percentile(agg.samples, 50)
            agg.p95 = _percentile(agg.samples, 95)
            agg.p99 = _percentile(agg.samples, 99)
            if agg.min == float("inf"):
                agg.min = 0.0
            if agg.max == float("-inf"):
                agg.max = 0.0
            agg.samples = []
            result.append(agg)
        return result


class RingBuffer:
    """Manages per-metric ring buffers, keyed by 'service|metric'."""

    def __init__(self, slots: int = 120, window_dur: float = 30.0):
        self._lock = threading.RLock()
        self._rings: dict[str, MetricRing] = {}
        self._slots = slots
        self._window_dur = window_dur

    def record(self, metric_name: str, service_name: str, value: float, at: datetime | None = None) -> None:
        key = f"{service_name}|{metric_name}"
        ts = at.timestamp() if at else _time.time()

        with self._lock:
            ring = self._rings.get(key)
            if ring is None:
                ring = MetricRing(metric_name, service_name, self._slots, self._window_dur)
                self._rings[key] = ring

        ring.record(value, ts)

    def query_recent(self, metric_name: str, service_name: str, window_count: int = 10) -> list[WindowAgg]:
        key = f"{service_name}|{metric_name}"
        with self._lock:
            ring = self._rings.get(key)
        if ring is None:
            return []
        return ring.windows(window_count)

    def all_keys(self) -> list[str]:
        with self._lock:
            return list(self._rings.keys())

    def metric_count(self) -> int:
        with self._lock:
            return len(self._rings)
